File: tetris3d.js
// tetris3d.js - 3D Тетрис на JavaScript (Node.js)
const fs = require('fs');
const keypress = require('keypress');
const readline = require('readline');

const WIDTH = 6;
const DEPTH = 6;
const HEIGHT = 6;
const EMPTY = 0;
const RECORD_FILE = 'tetris3d_record.json';

// Фигуры
const SHAPES = {
    I: [[0,0,0],[1,0,0],[2,0,0],[3,0,0]],
    O: [[0,0,0],[1,0,0],[0,1,0],[1,1,0]],
    T: [[0,0,0],[1,0,0],[2,0,0],[1,1,0]],
    L: [[0,0,0],[1,0,0],[2,0,0],[2,1,0]],
    J: [[0,0,0],[1,0,0],[2,0,0],[0,1,0]],
    S: [[1,0,0],[2,0,0],[0,1,0],[1,1,0]],
    Z: [[0,0,0],[1,0,0],[1,1,0],[2,1,0]],
};

class Tetris3D {
    constructor() {
        this.field = Array.from({ length: HEIGHT }, () =>
            Array.from({ length: WIDTH }, () => Array(DEPTH).fill(EMPTY))
        );
        this.score = 0;
        this.speed = 1;
        this.gameOver = false;
        this.paused = false;
        this.running = true;
        this.record = this.loadRecord();
        this.currentShape = null;
        this.shapePos = [0,0,0];
        this.dropTimer = Date.now();
        this.lock = false;
    }

    loadRecord() {
        try {
            const data = fs.readFileSync(RECORD_FILE, 'utf8');
            return JSON.parse(data).record || 0;
        } catch { return 0; }
    }

    saveRecord() {
        fs.writeFileSync(RECORD_FILE, JSON.stringify({ record: this.record }));
    }

    randomShape() {
        const names = Object.keys(SHAPES);
        const name = names[Math.floor(Math.random() * names.length)];
        return SHAPES[name].map(p => [...p]);
    }

    spawnShape() {
        const shape = this.randomShape();
        const xs = shape.map(p => p[0]);
        const zs = shape.map(p => p[1]);
        const cx = Math.floor((Math.max(...xs) + Math.min(...xs)) / 2);
        const cz = Math.floor((Math.max(...zs) + Math.min(...zs)) / 2);
        const startX = Math.floor(WIDTH / 2) - cx;
        const startZ = Math.floor(DEPTH / 2) - cz;
        const startY = HEIGHT - 1;
        // Проверка
        for (const [dx, dz, dy] of shape) {
            const x = startX + dx, z = startZ + dz, y = startY + dy;
            if (x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT || this.field[y][x][z] !== EMPTY) {
                this.gameOver = true;
                if (this.score > this.record) {
                    this.record = this.score;
                    this.saveRecord();
                }
                return false;
            }
        }
        this.currentShape = shape;
        this.shapePos = [startX, startZ, startY];
        return true;
    }

    collides(shape, pos) {
        for (const [dx, dz, dy] of shape) {
            const x = pos[0] + dx, z = pos[1] + dz, y = pos[2] + dy;
            if (x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT) return true;
            if (this.field[y][x][z] !== EMPTY) return true;
        }
        return false;
    }

    lockShape() {
        for (const [dx, dz, dy] of this.currentShape) {
            const x = this.shapePos[0] + dx, z = this.shapePos[1] + dz, y = this.shapePos[2] + dy;
            this.field[y][x][z] = 1;
        }
        // Удаление слоёв
        let cleared = 0;
        for (let y = 0; y < HEIGHT; y++) {
            if (this.field[y].every(row => row.every(cell => cell !== EMPTY))) {
                for (let y2 = y; y2 > 0; y2--) {
                    this.field[y2] = this.field[y2-1].map(row => [...row]);
                }
                this.field[0] = Array.from({ length: WIDTH }, () => Array(DEPTH).fill(EMPTY));
                cleared++;
            }
        }
        if (cleared) {
            this.score += cleared;
            this.speed = 1 + Math.floor(this.score / 3);
        }
        if (!this.spawnShape()) this.gameOver = true;
    }

    move(dx, dz, dy) {
        const newPos = [this.shapePos[0] + dx, this.shapePos[1] + dz, this.shapePos[2] + dy];
        if (!this.collides(this.currentShape, newPos)) {
            this.shapePos = newPos;
            return true;
        }
        return false;
    }

    rotate() {
        const newShape = this.currentShape.map(([x, z, y]) => [-z, x, y]);
        if (!this.collides(newShape, this.shapePos)) {
            this.currentShape = newShape;
            return true;
        }
        return false;
    }

    drop() {
        while (this.move(0, 0, -1)) {}
        this.lockShape();
    }

    update() {
        if (this.gameOver || this.paused) return;
        if (Date.now() - this.dropTimer >= 1000 / this.speed) {
            if (!this.move(0, 0, -1)) {
                this.lockShape();
            }
            this.dropTimer = Date.now();
        }
    }

    getFieldWithShape() {
        const fieldCopy = this.field.map(plane => plane.map(row => [...row]));
        if (this.currentShape) {
            for (const [dx, dz, dy] of this.currentShape) {
                const x = this.shapePos[0] + dx, z = this.shapePos[1] + dz, y = this.shapePos[2] + dy;
                if (x >= 0 && x < WIDTH && z >= 0 && z < DEPTH && y >= 0 && y < HEIGHT) {
                    fieldCopy[y][x][z] = 2;
                }
            }
        }
        return fieldCopy;
    }

    draw() {
        console.clear();
        const field = this.getFieldWithShape();
        console.log('═'.repeat(WIDTH * 2 + 10));
        console.log(`  Счёт: ${this.score}   Рекорд: ${this.record}   Скорость: ${this.speed}`);
        console.log('═'.repeat(WIDTH * 2 + 10));
        for (let y = HEIGHT - 1; y >= 0; y--) {
            process.stdout.write(`  Y=${y}  `);
            for (let z = 0; z < DEPTH; z++) {
                for (let x = 0; x < WIDTH; x++) {
                    const val = field[y][x][z];
                    if (val === 2) process.stdout.write('\x1b[32m█ \x1b[0m');
                    else if (val === 1) process.stdout.write('\x1b[36m█ \x1b[0m');
                    else process.stdout.write('· ');
                }
                process.stdout.write('  ');
            }
            console.log();
        }
        console.log('═'.repeat(WIDTH * 2 + 10));
        const status = this.paused ? 'ПАУЗА' : this.gameOver ? 'ИГРА ОКОНЧЕНА' : 'ИГРА';
        console.log(`  ${status}  |  ←/→ - X  |  W/S - Z  |  Пробел - вращение  |  ↓ - падение`);
        console.log(`  P - пауза  |  R - рестарт  |  Q - выход`);
    }

    reset() {
        this.field = Array.from({ length: HEIGHT }, () =>
            Array.from({ length: WIDTH }, () => Array(DEPTH).fill(EMPTY))
        );
        this.score = 0;
        this.speed = 1;
        this.gameOver = false;
        this.paused = false;
        this.spawnShape();
    }

    handleKey(ch, key) {
        if (!key) return;
        if (this.lock) return;
        this.lock = true;
        const action = () => {
            if (key.name === 'left') { if (!this.gameOver && !this.paused) this.move(-1,0,0); }
            else if (key.name === 'right') { if (!this.gameOver && !this.paused) this.move(1,0,0); }
            else if (key.name === 'w') { if (!this.gameOver && !this.paused) this.move(0,-1,0); }
            else if (key.name === 's') { if (!this.gameOver && !this.paused) this.move(0,1,0); }
            else if (key.name === 'space') { if (!this.gameOver && !this.paused) this.rotate(); }
            else if (key.name === 'down') { if (!this.gameOver && !this.paused) this.drop(); }
            else if (key.name === 'p') { this.paused = !this.paused; }
            else if (key.name === 'r') { this.reset(); }
            else if (key.name === 'q') { this.running = false; process.stdin.pause(); process.exit(0); }
            this.lock = false;
        };
        setTimeout(action, 10);
    }

    run() {
        this.spawnShape();
        keypress(process.stdin);
        process.stdin.setRawMode(true);
        process.stdin.resume();
        process.stdin.on('keypress', (ch, key) => this.handleKey(ch, key));

        const gameLoop = () => {
            if (!this.running) return;
            this.update();
            this.draw();
            setTimeout(gameLoop, 1000 / 30);
        };
        gameLoop();
    }
}

const game = new Tetris3D();
game.run();

File: tetris3d.py

---

## 💻 Код на 7 языках

### 1. Python – `tetris3d.py`

```python
#!/usr/bin/env python3
# tetris3d.py - 3D Тетрис на Python

import os
import sys
import time
import random
import json
import threading
from copy import deepcopy
from colorama import init, Fore, Style

init(autoreset=True)

# Константы
WIDTH = 6      # X
DEPTH = 6      # Z
HEIGHT = 6     # Y
EMPTY = 0
FIELD = [[[EMPTY for _ in range(DEPTH)] for _ in range(WIDTH)] for _ in range(HEIGHT)]
RECORD_FILE = 'tetris3d_record.json'

# Фигуры: список координат (x, z, y) относительных (локальных)
SHAPES = {
    'I': [(0,0,0), (1,0,0), (2,0,0), (3,0,0)],
    'O': [(0,0,0), (1,0,0), (0,1,0), (1,1,0)],
    'T': [(0,0,0), (1,0,0), (2,0,0), (1,1,0)],
    'L': [(0,0,0), (1,0,0), (2,0,0), (2,1,0)],
    'J': [(0,0,0), (1,0,0), (2,0,0), (0,1,0)],
    'S': [(1,0,0), (2,0,0), (0,1,0), (1,1,0)],
    'Z': [(0,0,0), (1,0,0), (1,1,0), (2,1,0)],
}

class Tetris3D:
    def __init__(self):
        self.field = deepcopy(FIELD)
        self.score = 0
        self.speed = 1
        self.game_over = False
        self.paused = False
        self.running = True
        self.record = self.load_record()
        self.current_shape = None
        self.shape_pos = [0, 0, 0]  # x, z, y
        self.drop_timer = time.time()
        self.lock = threading.Lock()

    def load_record(self):
        try:
            with open(RECORD_FILE, 'r') as f:
                return json.load(f).get('record', 0)
        except:
            return 0

    def save_record(self):
        with open(RECORD_FILE, 'w') as f:
            json.dump({'record': self.record}, f)

    def random_shape(self):
        name = random.choice(list(SHAPES.keys()))
        shape = [(x, z, y) for x, z, y in SHAPES[name]]
        return shape

    def spawn_shape(self):
        shape = self.random_shape()
        # Центрирование по X и Z
        xs = [p[0] for p in shape]
        zs = [p[1] for p in shape]
        cx = (max(xs) + min(xs)) // 2
        cz = (max(zs) + min(zs)) // 2
        start_x = WIDTH // 2 - cx
        start_z = DEPTH // 2 - cz
        start_y = HEIGHT - 1  # сверху
        # Проверка возможности спавна
        for dx, dz, dy in shape:
            x = start_x + dx
            z = start_z + dz
            y = start_y + dy
            if not (0 <= x < WIDTH and 0 <= z < DEPTH and 0 <= y < HEIGHT) or self.field[y][x][z] != EMPTY:
                self.game_over = True
                if self.score > self.record:
                    self.record = self.score
                    self.save_record()
                return False
        self.current_shape = shape
        self.shape_pos = [start_x, start_z, start_y]
        return True

    def collides(self, shape, pos):
        for dx, dz, dy in shape:
            x = pos[0] + dx
            z = pos[1] + dz
            y = pos[2] + dy
            if not (0 <= x < WIDTH and 0 <= z < DEPTH and 0 <= y < HEIGHT):
                return True
            if self.field[y][x][z] != EMPTY:
                return True
        return False

    def lock_shape(self):
        for dx, dz, dy in self.current_shape:
            x = self.shape_pos[0] + dx
            z = self.shape_pos[1] + dz
            y = self.shape_pos[2] + dy
            self.field[y][x][z] = 1  # блок
        # Проверка заполненных слоёв
        cleared = 0
        for y in range(HEIGHT):
            if all(self.field[y][x][z] != EMPTY for x in range(WIDTH) for z in range(DEPTH)):
                # Удалить слой
                for y2 in range(y, 0, -1):
                    self.field[y2] = self.field[y2-1]
                self.field[0] = [[EMPTY for _ in range(DEPTH)] for _ in range(WIDTH)]
                cleared += 1
        if cleared > 0:
            self.score += cleared
            self.speed = 1 + self.score // 3
        # Спавн новой фигуры
        if not self.spawn_shape():
            self.game_over = True

    def move(self, dx, dz, dy):
        new_pos = [self.shape_pos[0] + dx, self.shape_pos[1] + dz, self.shape_pos[2] + dy]
        if not self.collides(self.current_shape, new_pos):
            self.shape_pos = new_pos
            return True
        return False

    def rotate(self):
        # Вращение вокруг оси Y (меняем x и z)
        new_shape = []
        for x, z, y in self.current_shape:
            new_shape.append((-z, x, y))  # поворот на 90° по часовой
        if not self.collides(new_shape, self.shape_pos):
            self.current_shape = new_shape
            return True
        return False

    def drop(self):
        while self.move(0, 0, -1):
            pass
        self.lock_shape()

    def update(self):
        if self.game_over or self.paused:
            return
        # Автопадание
        if time.time() - self.drop_timer >= 1.0 / self.speed:
            if not self.move(0, 0, -1):
                self.lock_shape()
            self.drop_timer = time.time()

    def get_field_with_shape(self):
        field_copy = deepcopy(self.field)
        if self.current_shape:
            for dx, dz, dy in self.current_shape:
                x = self.shape_pos[0] + dx
                z = self.shape_pos[1] + dz
                y = self.shape_pos[2] + dy
                if 0 <= x < WIDTH and 0 <= z < DEPTH and 0 <= y < HEIGHT:
                    field_copy[y][x][z] = 2  # фигура
        return field_copy

    def draw(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print('═' * (WIDTH * 2 + 10))
        print(f'  Счёт: {self.score}   Рекорд: {self.record}   Скорость: {self.speed}')
        print('═' * (WIDTH * 2 + 10))
        # Отображение слоёв (сверху вниз)
        for y in range(HEIGHT-1, -1, -1):
            print(f'  Y={y}  ', end='')
            for z in range(DEPTH):
                for x in range(WIDTH):
                    val = self.field[y][x][z]
                    if val == 2:
                        print(Fore.GREEN + '█' + Style.RESET_ALL, end=' ')
                    elif val == 1:
                        print(Fore.CYAN + '█' + Style.RESET_ALL, end=' ')
                    else:
                        print('·', end=' ')
                print('  ', end='')
            print()
        # Отображение текущей фигуры (проекция)
        print('═' * (WIDTH * 2 + 10))
        status = "ПАУЗА" if self.paused else ("ИГРА ОКОНЧЕНА" if self.game_over else "ИГРА")
        print(f'  {status}  |  ←/→ - X  |  W/S - Z  |  Пробел - вращение  |  ↓ - падение')
        print(f'  P - пауза  |  R - рестарт  |  Q - выход')

    def reset(self):
        self.field = deepcopy(FIELD)
        self.score = 0
        self.speed = 1
        self.game_over = False
        self.paused = False
        self.spawn_shape()

    def handle_input(self):
        while self.running:
            try:
                if keyboard.is_pressed('left'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.move(-1, 0, 0)
                elif keyboard.is_pressed('right'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.move(1, 0, 0)
                elif keyboard.is_pressed('w'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.move(0, -1, 0)
                elif keyboard.is_pressed('s'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.move(0, 1, 0)
                elif keyboard.is_pressed('space'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.rotate()
                elif keyboard.is_pressed('down'):
                    with self.lock:
                        if not self.game_over and not self.paused:
                            self.drop()
                elif keyboard.is_pressed('p'):
                    with self.lock:
                        self.paused = not self.paused
                    time.sleep(0.2)
                elif keyboard.is_pressed('r'):
                    with self.lock:
                        self.reset()
                    time.sleep(0.2)
                elif keyboard.is_pressed('q'):
                    self.running = False
                    break
            except:
                pass
            time.sleep(0.05)

    def run(self):
        # Начальный спавн
        self.spawn_shape()
        input_thread = threading.Thread(target=self.handle_input, daemon=True)
        input_thread.start()

        last_update = time.time()
        while self.running:
            now = time.time()
            if now - last_update >= 1.0 / 30:  # 30 FPS
                with self.lock:
                    self.update()
                    self.draw()
                last_update = now
            time.sleep(0.02)

if __name__ == "__main__":
    game = Tetris3D()
    try:
        game.run()
    except KeyboardInterrupt:
        print("\nВыход...")
